# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: Matthew Bilodeau


![LegoSchema](Lego_Schema.PNG)

## Overview

  Rebrickable is a website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  (https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/)




In [3]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [4]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

In [5]:
inventories.head(), inventories.info(), inventories.describe(), inventory_sets.head(), inventory_sets.info(), inventory_sets.describe,
sets.head(), sets.info(), sets.describe, themes.head(), themes.info(), themes.describe, inventory_minifigs.head(), inventory_minifigs.info(),   
minifigs.head(), minifigs.describe(), minifigs.info(), inventory_parts.head(), inventory_parts.info(), inventory_parts.describe(),
part_categories.head(), part_categories.info(), part_categories.describe(), parts.head(), parts.info(), parts.describe, 
colors.head(), colors.info(), colors.describe(), part_relationships.head(), part_relationships.info(), part_relationships.describe(),
elements.head(), elements.info(), elements.describe() 


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31587 entries, 0 to 31586
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       31587 non-null  int64 
 1   version  31587 non-null  int64 
 2   set_num  31587 non-null  object
dtypes: int64(2), object(1)
memory usage: 740.4+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3413 entries, 0 to 3412
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   inventory_id  3413 non-null   int64 
 1   set_num       3413 non-null   object
 2   quantity      3413 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 80.1+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18576 entries, 0 to 18575
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   set_num    18576 non-null  object
 1   name       18576 non-null  object
 2   year       18576 no

(  element_id  part_num  color_id
 0    6300211  67906c01        14
 1    4566309      2564         0
 2    4275423     53657      1004
 3    6194308     92926        71
 4    6229123     26561         4,
 None,
            color_id
 count  62684.000000
 mean     608.309983
 std     2210.989312
 min       -1.000000
 25%       14.000000
 50%       30.000000
 75%      148.000000
 max     9999.000000)

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [6]:
def data_count(x): 
    return int(len(x))

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [7]:
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [8]:
Q2 = round(sets['num_parts'].mean())

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [9]:
Q3 = int(sets['num_parts'].median())


You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [10]:
Q4 = sets['num_parts'].max()

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [11]:
Q5 = sets[sets['num_parts'] == Q4]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [12]:

def select_set_row(x):
  y =  sets.loc[sets['set_num'] == x]
  return y 
select_set_row('31203-1')

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [13]:
def select_set_numparts(x):
   
    y = sets.loc[sets['num_parts'] ==x]
    y = y.sort_values(['year','set_num'])
    return y 



**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [14]:
select_set_numparts(0)


set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [15]:
Q8A = int(sets['year'].min())


In [16]:
Q8B = int(sets['year'].max())

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [17]:
def select_set_year(x): 
    y = sets.loc[sets['year'] ==x]
    y.sort_values(['set_num'])
    return y 
select_set_year(1960)

set_num                         name  year  theme_id  num_parts
2485    237-1                Number Bricks  1960       371         50
2491    238-1  Lego System Idea Book no. 1  1960       497          0
11098  700K-1        Kindergarten LEGO Set  1960       513        476

**Q10:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [18]:

theme_by_year_df = sets.merge(themes,left_on= 'theme_id', right_on = 'id').sort_values([ 'id'])
def theme_by_year(x): 
    y = theme_by_year_df.loc[theme_by_year_df['year'] ==x]
    output = y[['id', 'name_y']].rename( columns = {'name_y' : 'name_themes'}).drop_duplicates(subset= ['id']).sort_values(
        by = 'id').reset_index(drop = True)
    return output

theme_by_year(1978)


id               name_themes
0     4            Expert Builder
1    67              Classic Town
2   189            Classic Castle
3   227                 Christmas
4   235                      4.5V
5   363                      Boat
6   364  Building Set with People
7   404                Hobby Sets
8   405                 Homemaker
9   412                   Airport
10  453                   Technic
11  469                 Basic Set
12  473              Supplemental
13  501                      Gear
14  505                 Basic Set
15  632                      Town

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [19]:
theme_by_year(1999)

id            name_themes
0     1                Technic
1     3            Competition
2    16             RoboRiders
3    20        Throwbot Slizer
4    22                Creator
5    50                   Town
6    87           Extreme Team
7    92                  Res-Q
8    93             Space Port
9    94               Town Jr.
10  106            City Center
11  133        Ice Planet 2002
12  134             Insectoids
13  140        Space Police II
14  142                Spyrius
15  144                    UFO
16  158              Star Wars
17  207                 Advent
18  227              Christmas
19  236                     9V
20  260                    RCX
21  261              Star Wars
22  278             Model Team
23  297                 Desert
24  299                 Jungle
25  311             Aquasharks
26  318               Belville
27  434                  Ninja
28  440                  Primo
29  442           Rock Raiders
30  443          Service Packs
31  453                Technic
32  469              Basic Set
33  470          Classic Basic
34  471                Ferries
35  480                   Znap
36  497                  Books
37  501                   Gear
38  504                  Duplo
39  505              Basic Set
40  507  Educational and Dacta
41  516      Duplo and Explore
42  520                    RCX
43  524          Service Packs
44  525                Technic
45  526            Soft Bricks
46  531                   eLAB
47  533                   Town
48  559            Value Packs
49  632                   Town
50  635        Winnie the Pooh
51  658                   Dino
52  660  Little Forest Friends
53  668                  Scala

**Q11:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [20]:
def theme_by_name(x):
    y = theme_by_year_df.loc[theme_by_year_df['name_y'] ==x].rename(columns={'name_x' :'name_sets', 
                                                                             'name_y' :'name_themes'}).drop(['id','parent_id'], axis = 1)
    z = y[['set_num', 'name_sets', 'year', 'num_parts','theme_id', 'name_themes']].sort_values(by = 'set_num').reset_index(drop = True)
    return z



**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [21]:
theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [22]:
theme_by_year_df = sets.merge(themes,left_on= 'theme_id', right_on = 'id').sort_values('year')
def theme_by_setnum(x): 
    y = theme_by_year_df.loc[theme_by_year_df['set_num'] ==x]
    z = str(y.iloc[0]['name_y'])
    return z 

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [23]:

print(theme_by_setnum('75810-1'))


Stranger Things


**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [24]:
m_falcon_largest = sets.loc[sets['name'] == 'Millennium Falcon']
m_falcon_largest = m_falcon_largest[m_falcon_largest['num_parts'] == m_falcon_largest['num_parts'].max()]
m_falcon_largest

set_num               name  year  theme_id  num_parts
12884  75192-1  Millennium Falcon  2017       171       7541

**Q13:** 
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [25]:

m_falcon_largest_inventories = m_falcon_largest.merge(inventories)
m_falcon_largest_inventories = m_falcon_largest_inventories.merge(inventory_minifigs, left_on = 'id',
                                                                  right_on = 'inventory_id').drop(['name','year',
                                                                                                   'num_parts','theme_id',
                                                                                                   'num_parts']
                                                                                                  ,axis = 1)
m_falcon_minifigs =m_falcon_largest_inventories.merge(minifigs)
Q13 = len(m_falcon_minifigs)
m_falcon_minifigs


set_num     id  version  inventory_id     fig_num  quantity  \
0  75192-1  19670        1         19670  fig-001714         1   
1  75192-1  19670        1         19670  fig-001810         1   
2  75192-1  19670        1         19670  fig-002057         1   
3  75192-1  19670        1         19670  fig-002514         1   
4  75192-1  19670        1         19670  fig-002544         1   
5  75192-1  19670        1         19670  fig-002549         1   
6  75192-1  19670        1         19670  fig-002550         1   
7  75192-1  19670        1         19670  fig-002551         1   

                                                name  num_parts  
0                              Chewbacca, Dark Brown          3  
1                             Finn, Black Undershirt          4  
2                                 Rey, Dark Tan Robe          4  
3    C-3PO, Pearl Gold, Colorful Wires, Printed Legs          3  
4                                               BB-8          2  
5  Princess Leia, Hoth Outfit, White Jacket, Prin...          4  
6  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...          4  
7                               Han Solo, Old, Angry          4

In [26]:
theme_by_year_df = sets.merge(themes,left_on= 'theme_id', right_on = 'id').sort_values('year')
def theme_by_setnum(x): 
    y = theme_by_year_df.loc[theme_by_year_df['set_num'] ==x]
    return y
theme_by_setnum('75192-1')

set_num             name_x  year  theme_id  num_parts   id  \
3199  75192-1  Millennium Falcon  2017       171       7541  171   

                         name_y  parent_id  
3199  Ultimate Collector Series      158.0

**Q14:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [27]:

get_minifigs_df = sets.merge(inventories).drop(['name','year','num_parts'],axis =1)
get_minifigs_df = get_minifigs_df.merge(inventory_minifigs, left_on = 'id', right_on = 'inventory_id')
get_minifigs_df = get_minifigs_df.merge(minifigs, on = 'fig_num').reset_index(drop = True)
def get_minifigs(x,y): 
    z = get_minifigs_df.loc[(get_minifigs_df['set_num'] == x) & (get_minifigs_df['version'] == y)].sort_values(
        by = 'set_num').drop(['id', 'version', 'inventory_id','quantity'],axis =1).rename(columns = {
            'name': 'fig_name',
            'num_parts': 'fig_num_parts'    
        }).drop(columns={'theme_id'}).sort_values(by = 'fig_num').reset_index(drop = True)
    z = z[['set_num', 'fig_num', 'fig_name', 'fig_num_parts']]
    return z 
get_minifigs_df.columns

Index(['set_num', 'theme_id', 'id', 'version', 'inventory_id', 'fig_num',
       'quantity', 'name', 'num_parts'],
      dtype='object')

**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [28]:
Q15 = get_minifigs('75192-1', 1)

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [29]:
get_minifigs('10217-1',2)


set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q16:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [30]:

get_minifigs_df = sets.merge(inventories).drop(['year','num_parts'],axis =1).rename(columns = {'name': 'name_sets'}).reset_index(drop = True)
get_minifigs_df = get_minifigs_df.merge(inventory_minifigs, left_on = 'id', right_on = 'inventory_id').reset_index(drop = True)
get_minifigs_df = get_minifigs_df.merge(minifigs, on = 'fig_num').reset_index(drop = True)
def minifigs_from_themes(x): 
    y = get_minifigs_df.loc[(get_minifigs_df['theme_id'] == x)].reset_index(drop = True)
    z = y[['fig_num','name']].drop_duplicates().rename(columns = {'name' :'fig_name'}).sort_values(['fig_num']).reset_index(drop = True)
    return z


**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [31]:
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q17:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [32]:
set_minifigs_df = sets.merge(inventories).drop(['year','num_parts'],axis =1).rename(columns = {'name': 'name_sets'})
set_minifigs_df = get_minifigs_df.merge(inventory_minifigs, left_on = 'id', right_on = 'inventory_id')
set_minifigs_df = get_minifigs_df.merge(minifigs, on = 'fig_num').reset_index(drop = True).rename(columns = {'name_y': 'fig_name'})


def sets_from_minifig(x): 
   a = set_minifigs_df.loc[(set_minifigs_df['fig_num'] == x)].sort_values(['fig_num'])
   b = a[['set_num', 'version', 'name_sets', 'fig_num', 'fig_name']].reset_index(drop = True).sort_values(by = 'set_num')
   return b
sets_from_minifig('fig-001039')




set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [33]:
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q18:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [34]:
sets = sets.rename({'name': 'name_sets'}, axis = 1)
colors = colors.rename({'id': 'color_id', 'name': 'name_color'},axis = 1)
parts = parts.rename({'name': 'name_part'}, axis = 1)
themes = themes.rename({'name': 'name_themes'}, axis = 1)
set_parts_df = sets.merge(inventories).merge(inventory_parts,
                                             left_on = 'id',
                                             right_on= 'inventory_id').merge(colors).merge(parts).merge(themes,
                                                    left_on= 'theme_id', right_on = 'id').reset_index(drop = True)
set_parts_df = set_parts_df[['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 'name_part', 'name_color', 'quantity', 'is_spare']].reset_index(drop = True)

def set_parts(x, y): 
    z = set_parts_df.loc[(set_parts_df['set_num'] == x) & (set_parts_df['version'] == y)].sort_values(by = ['part_num', 'name_color']).reset_index(drop = True)
    return z
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [35]:
set_parts('75192-1',1)

set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
725    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
726    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
727    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
728    99781                              Bracket 1 x 2 - 1 x 2   
729    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
0     Dark Bluish Gray         3        f  
1    Light Bluish Gray        12        f  
2     Dark Bluish Gray         5        f  
3                White        22        f  
4                  Tan         3        f  
..                 ...       ...      ...  
725  Light Bluish Gray        70        f  
726              Black         6        f  
727   Dark Bluish Gray        25        f  
728  Light Bluish Gray         2        f  
729  Light Bluish Gray         3        f  

[730 rows x 10 columns]